articles/realtime.md

```diff
@@ -0,0 +1,259 @@
+---
+layout: default
+title: ROS 2.0 in Realtime
+permalink: articles/realtime.html
+abstract: This article describes real-time computing requirements and how ROS 2 will fulfill them.
+published: false
+author: Jackie Kay
+---
+
+* This will become a table of contents (this text will be scraped).
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+Robotic systems need to be responsive.
+In mission critical applications, a delay of less than a millisecond in the system can cause a catastrophic failure.
+For ROS 2.0 to capture the needs of the robotics community, the core software components must not interfere with the requirements of real-time computing.
+
+# Definition of Real-time Computing
+
+Real-time software guarantees correct computation at the correct time.
+Hard real-time software systems have a set of strict deadlines, and missing a deadline is considered a failure.
+Soft real-time systems degrade their quality of service if a deadline is missed.
+
+Real-time computer systems are often associated with low-latency systems.
+Many applications of real-time computing are also low-latency applications (for example, automated piloting systems must be reactive to sudden changes in the environment).
+However, it is generally agreed upon that a real-time system is not defined by low latency, but by a deterministic schedule: it must be guaranteed that the system finishes a certain task by a certain time.
+Therefore, it is important that the latency in the system be measurable and a maximum allowable latency for tasks be set.
+
+A more useful metric in evaluating the "hardness" of a real-time system is jitter, which is defined as the variable deviation between a task's deadline and its actual time of completion.
+A more jittery system is less deterministic, less predictable, and less real-time.
+Though in practice it is impossible to completely eliminate jitter from a real-time system, it is a worthy goal to determine a hard upper bound for jitter.
+
```

Comment:
Before the next section, it could be mentioned that a system that meets real-time requirements needs both an operating system (kernel, scheduler) **and** user code capable of delivering deterministic execution.

It might be out of scope, but a comment on RTOS alternatives could be interesting. In the Linux world, it's worth mentioning:

* Linux kernel patches like [PREEMPT_RT ](https://rt.wiki.kernel.org/index.php/Main_Page) which '...allows nearly all of the kernel to be preempted'.

* Hard real-time co-kernels (also called hypervisors) that cooperate with the Linux kernel like [Xenomai](http://xenomai.org/) and [RTAI](https://www.rtai.org/). They treat Linux as the idle task of the real-time scheduler.

There are also RTOS for mission-critical and safety-critical systems, like [QNX Neutrino](http://www.qnx.com/products/neutrino-rtos/neutrino-rtos.html) (fully POSIX compliant) and [VxWorks](http://www.windriver.com/vxworks/reinvented/?gclid=COrv7YXrtcYCFWv3wgodfEcKqA) (POSIX compliant through compatibility layer) that are certified to IEC 61508 SIL 3.